File: 08_binary_search/08_07_search_in_rotated_sorted_array_I/search_in_rotated_sorted_array_I.py
def search(nums, target):
    low, high = 0, len(nums) - 1
    while low <= high:
        mid = low + (high - low) // 2
        if nums[mid] == target:
            return mid
        elif nums[low] <= nums[mid]:
            if nums[low] <= target <= nums[mid]:
                high = mid - 1
            else:
                low = mid + 1
        else:
            if nums[mid] <= target <= nums[high]:
                low = mid + 1
            else:
                high = mid - 1
    return -1

# Test cases
nums1 = [4, 5, 6, 7, 0, 1, 2]
target1 = 0
print(f"Test Case 1 - Array: {nums1}")
print(f"Target: {target1}")
print(f"Index: {search(nums1, target1)}")

nums2 = [4, 5, 6, 7, 0, 1, 2]
target2 = 3
print(f"\nTest Case 2 - Array: {nums2}")
print(f"Target: {target2}")
print(f"Index: {search(nums2, target2)}")

nums3 = [1]
target3 = 0
print(f"\nTest Case 3 - Array: {nums3}")
print(f"Target: {target3}")
print(f"Index: {search(nums3, target3)}")

nums4 = [4, 5, 6, 7, 8, 1, 2, 3]
target4 = 8
print(f"\nTest Case 4 - Array: {nums4}")
print(f"Target: {target4}")
print(f"Index: {search(nums4, target4)}")

File: 08_binary_search/08_07_search_in_rotated_sorted_array_I/search_in_rotated_sorted_array_I.js
function search(nums, target) {
    let low = 0, high = nums.length - 1;
    while (low <= high) {
        let mid = low + Math.floor((high - low) / 2);
        if (nums[mid] === target) {
            return mid;
        } else if (nums[low] <= nums[mid]) {
            if (nums[low] <= target && target <= nums[mid]) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        } else {
            if (nums[mid] <= target && target <= nums[high]) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
    }
    return -1;
}

// Test cases
let nums1 = [4, 5, 6, 7, 0, 1, 2];
let target1 = 0;
console.log(`Test Case 1 - Array: [${nums1}]`);
console.log(`Target: ${target1}`);
console.log(`Index: ${search(nums1, target1)}`);

let nums2 = [4, 5, 6, 7, 0, 1, 2];
let target2 = 3;
console.log(`\nTest Case 2 - Array: [${nums2}]`);
console.log(`Target: ${target2}`);
console.log(`Index: ${search(nums2, target2)}`);

let nums3 = [1];
let target3 = 0;
console.log(`\nTest Case 3 - Array: [${nums3}]`);
console.log(`Target: ${target3}`);
console.log(`Index: ${search(nums3, target3)}`);

let nums4 = [4, 5, 6, 7, 8, 1, 2, 3];
let target4 = 8;
console.log(`\nTest Case 4 - Array: [${nums4}]`);
console.log(`Target: ${target4}`);
console.log(`Index: ${search(nums4, target4)}`);
